src/backend/distributed/README.md

```diff
@@ -0,0 +1,2079 @@
+
+The principle behind `citus.max_shared_pool_size` is that under high concurrency (all client connections used) it converges to each process having 1 connection per node. To do so, we distinguish between “optional” and “required” connections. When the executor asks the connection management layer for a connection, the first connection to a node is always required, and other connections are optional. If all connection slots are in use, the connection manager blocks until one is available when asking for a required connection, or returns NULL when asking for an optional connection. That signals to the executor that it cannot currently expand its pool. It may try again later. Most Citus code paths are tweaked to be able to complete their operation with 1 connection per node, and use local execution for local shards. 
+
+Note that `citus.max_shared_pool_size` can only control the number of outgoing connections on a single node. When there are many nodes, the number of possible inbound internal connections is the sum of the `citus.max_shared_pool_size` on all other nodes. To ensure this does not exceed max_connections, we recommend that `sum(citus.max_client_connections) < max_connections`.  
+
+# Transactions (2PC) 
+
+Citus uses the transaction callbacks in PostgreSQL for pre-commit, post-commit, and abort to implement distributed transactions. In general, distributed transactions comprise a transaction on the coordinator and one or more transactions on worker nodes. For transactions that only involve a single worker node, Citus delegates responsibility to the worker node. For transactions that involve multiple nodes, Citus uses two-phase commit for atomicity and implements distributed deadlock detection. 
+
+## Single-node transactions 
+
+Most multi-tenant and high-performance CRUD workloads only involve transactions that access a single worker node (or rather, a single shard group). For example, multi-tenant applications typically distribute and co-locate tables by tenant and transactions typically only involve a single tenant. When all statements in a transaction are routed to the same worker node, the coordinator simply sends commit/abort commands to that worker node from the commit/abort callbacks. In this case, the transaction is effectively delegated to that worker node. The worker node, by definition, provides the same guarantees as a single PostgreSQL server. 
+
+## Multi-node transactions 
+
```

Comment:
Not a doc suggestion, just a question... and I'm not sure it's well-formed. If my transaction reads from shard 1 on node 1 and writes to shard 2 on node 2, is that a multi-node transaction? How is it handled? Does this case fall under "No distributed transaction isolation"?

Reply:
It is not a multi-node transaction in the sense of a 2PC. We have an aptly named `Activate2PCIfModifyingTransactionExpandsToNewNode` function, which tracks whether we modify things on multiple nodes.

Isolation is not so important if all queries are single shard, because in read committed node every query is expected to get its own snapshot. We mainly differ from Postgres for multi-shard queries.